lib/compress/zstdmt_compress.c

```diff
@@ -1612,6 +1623,11 @@ static size_t ZSTDMT_createCompressionJob(ZSTDMT_CCtx* mtctx, size_t srcSize, ZS
     }
 
     if (!mtctx->jobReady) {
+        if (mtctx->cJobControl.canCreateJob && !mtctx->cJobControl.canCreateJob (mtctx->cJobControl.opaque)) {
+          DEBUGLOG(5, "ZSTDMT_createCompressionJob: canCreateJob returned false");
+          return 0;
```

Comment:
Expected behavior should be well documented here.
Is `canCreateJob` a blocking function ?
What's the impact of `return 0` ?
Will it cancel the job ? Will it make the user process enter a busy-loop ?

Reply:
I tried to document it properly in `zstd.h`, can you please take a look?
It's not a blocking function and you are right, it enters a busy-loop. For a compression running for 40s, I see the `canCreateJob returned false` message 850 times in log. Is it acceptable?

Reply:
This is indeed a concern. Busy loops must be avoided.

To avoid this problem, `zstd` currently ensures forward progress : it only returns if at least one byte is consumed or produced.
When nothing happened, typically as a consequence of having no thread available to start a new job, it blocks the calling thread, up to the moment that "something" happens, typically a job just produced some data to flush, and/or got freed, hence allowing to start a new job.

This requires a capability to block / sleep the thread,
and more importantly, it requires a capability to wake up,
right on time, when a thread finally becomes available.

This is not trivial, and is likely even more difficult when trying to achieve a similar goal outside of the thread pool.

Reply:
I see.

Note that it's for me quite difficult to adopt the job control to the existing `zstd` codebase.
Maybe a closer look from a developer can introduce a more elegant way?